### PR TITLE
odo describe component

### LINKED ISSUE #5661: `odo describe component`
/kind user-story

## User Story
- As an odo user
- I want to be able to get detailed information about my component
- So that 
   - I know what commands/odo features are supported by given devfile 
   - I know what are the parts that my application  (devfile) is composed of 

## Acceptance Criteria
- [ ] when user executes the `odo describe component` it should display detailed information about the devfile.yaml in the current directory
- [ ] when user executed the `odo describe component` with `--name` flag it should display detailed information about the component existing on the cluster. Note that in this case the information is not extracted from devfile, but rather from Kubernetes resources. Some information might not be available (like supported odo features), in this case it should show `unknown`
- [ ] `--namespace` is optional and can be used only with `--name` flag
- [ ] forwarded ports are displayed only when `odo dev` is currently running. the format is `<local-address>:<local-port> -> <container-name>:<container-port>` this request to implement https://github.com/redhat-developer/odo/issues/5676 first

/kind user-story

Json output for this command is part of https://github.com/redhat-developer/odo/issues/5660

## example 
```
$ odo describe component

Name: python
Display name: Python
Project Type: python
Language: python
Version: 1.0.0
Description: Python Stack with Python 3.7
Tags: Python, pip

Running in: Dev,Deploy

Forwarded ports:
   - localhost:40001 -> py-web:8080 

Supported odo features:
  - dev: true
  - deploy: false
  - debug: true

Container components:
 - py-web

Kubernetes components:
  - binding-mysql
```

```
$ odo describe component --name myspring

Name: myspring
Display name: 
Project Type: springboot
Language: java
Version: 1.0.0
Description: unknown
Tags: unknown

Supported odo features:
  - dev: unknown
  - deploy: unknown
  - debug: unknown

Running in: Dev,Deploy

Container components: unknown

Kubernetes components: unknown
```

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #5661 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

Doc preview: https://deploy-preview-5725--odo-docusaurus-preview.netlify.app/docs/3.0.0/command-reference/describe

### CODE REVIEW

File: pkg/odo/cli/describe/component.go [hunk 1]
```diff
@@ -139,11 +150,76 @@ func (o *ComponentOptions) describeDevfileComponent() (result api.Component, err
...
+	fmt.Println("Supported odo features:")
+	if cmp.DevfileData != nil && cmp.DevfileData.SupportedOdoFeatures != nil {
+		fmt.Printf(" - Dev: %v\n", cmp.DevfileData.SupportedOdoFeatures.Dev)
+		fmt.Printf(" - Deploy: %v\n", cmp.DevfileData.SupportedOdoFeatures.Deploy)
+		fmt.Printf(" - Debug: %v\n", cmp.DevfileData.SupportedOdoFeatures.Debug)
+	} else {
+		fmt.Printf(" - Dev: Unknown\n")
+		fmt.Printf(" - Deploy: Unknown\n")
+		fmt.Printf(" - Debug: Unknown\n")
+	}
```
Comment (valaparthvi) [inline pkg/odo/cli/describe/component.go:188]: Perhaps you can reintroduce [Describef](https://github.com/redhat-developer/odo/blob/v2/pkg/log/status.go#L338) with this PR.

```suggestion
	if cmp.DevfileData != nil {
		log.Describef("Name: ", cmp.DevfileData.Devfile.GetMetadata().Name)
		log.Describef("Display Name: ", cmp.DevfileData.Devfile.GetMetadata().DisplayName)
		log.Describef("Project Type: ", cmp.DevfileData.Devfile.GetMetadata().ProjectType)
		log.Describef("Language: ", cmp.DevfileData.Devfile.GetMetadata().Language)
		log.Describef("Version: ", cmp.DevfileData.Devfile.GetMetadata().Version)
		log.Describef("Description: ", cmp.DevfileData.Devfile.GetMetadata().Description)
		log.Describef("Tags: ", strings.Join(cmp.DevfileData.Devfile.GetMetadata().Tags, ", "))
		fmt.Println()
	}

	log.Describef("Running in: ", cmp.RunningIn.String())
	fmt.Println()

	if len(cmp.DevForwardedPorts) > 0 {
		log.Info("Forwarded ports:")
		for _, port := range cmp.DevForwardedPorts {
			log.Printf("%s:%d -> %s:%d", port.LocalAddress, port.LocalPort, port.ContainerName, port.ContainerPort)
		}
		fmt.Println()
	}

	log.Info("Supported odo features:")
	if cmp.DevfileData != nil && cmp.DevfileData.SupportedOdoFeatures != nil {
		log.Printf("Dev: %v", cmp.DevfileData.SupportedOdoFeatures.Dev)
		log.Printf("Deploy: %v", cmp.DevfileData.SupportedOdoFeatures.Deploy)
		log.Printf("Debug: %v", cmp.DevfileData.SupportedOdoFeatures.Debug)
	} else {
		log.Printf("Dev: Unknown")
		log.Printf("Deploy: Unknown")
		log.Printf("Debug: Unknown")
	}
```

<details>
<summary>**Outcome:**</summary>

Screenshot from 2022-05-11 00-45-09
</details>

File: pkg/odo/cli/describe/component.go [hunk 2]
```diff
@@ -139,11 +150,76 @@ func (o *ComponentOptions) describeDevfileComponent() (result api.Component, err
...
+		return err
+	}
+	return nil
 }
 
-func printHumanReadableOutput(component api.Component) {
-	// TODO(feloy) #5661
+func listComponentsNames(title string, devfileObj *parser.DevfileObj, typ v1alpha2.ComponentType) error {
+	if devfileObj == nil {
+		fmt.Printf("%s Unknown\n\n", title)
```
Comment (valaparthvi) [inline pkg/odo/cli/describe/component.go:205]:
```suggestion
		log.Describef(title, " Unknown")
```

File: pkg/odo/cli/describe/component.go [hunk 3]
```diff
@@ -139,11 +150,76 @@ func (o *ComponentOptions) describeDevfileComponent() (result api.Component, err
...
+	if err != nil {
+		return err
+	}
+	if len(containers) == 0 {
+		return nil
+	}
+	fmt.Println(title)
+	for _, container := range containers {
+		fmt.Printf(" - %s\n", container.Name)
+	}
```
Comment (valaparthvi) [inline pkg/odo/cli/describe/component.go:220]:
```suggestion
	log.Info(title)
	for _, container := range containers {
		log.Printf("%s", container.Name)
	}
```

**Review by valaparthvi — CHANGES_REQUESTED** [review]
The changes look good, I only have a few questions/suggestions.

**Review by valaparthvi — APPROVED** [review]
Thank you for working on those changes, Philippe!

**Review by rm3l — COMMENTED** [review]
LGTM, but I had a few small questions (might be enhancements for other PRs) :)

- In the "Running in" section, do we want to be more specific with the "Debug" Dev mode? Because the command displays "Dev" no matter if I run `odo dev` or `odo dev --debug`.
I found it a bit strange given that the "Supported odo features" section clearly lists "Debug" as a possible feature supported, but this is currently not listed in "Running in" when running in this mode.

- Also, I understood that the "Supported odo features" information is determined from the current Devfile and currently displays "Unknown" when running with `--name`. When running with the `--name` flag, if a mode (`Dev` and/or `Deploy`) is detected and listed in the "Running in" section, it means that this feature is supported by `odo`, no?
```
❯ odo describe component --name my-app             
Name: my-app
Display Name: Unknown
Project Type: quarkus
Language: Unknown
Version: Unknown
Description: Unknown
Tags: 

Running in: Dev, Deploy

Supported odo features:
 •  Dev: Unknown
 •  Deploy: Unknown
 •  Debug: Unknown
```

**Comment by feloy** [comment]
> LGTM, but I had a few small questions (might be enhancements for other PRs) :)
> 
> * In the "Running in" section, do we want to be more specific with the "Debug" Dev mode? Because the command displays "Dev" no matter if I run `odo dev` or `odo dev --debug`.
>   I found it a bit strange given that the "Supported odo features" section clearly lists "Debug" as a possible feature supported, but this is currently not listed in "Running in" when running in this mode.



> * Also, I understood that the "Supported odo features" information is determined from the current Devfile and currently displays "Unknown" when running with `--name`. When running with the `--name` flag, if a mode (`Dev` and/or `Deploy`) is detected and listed in the "Running in" section, it means that this feature is supported by `odo`, no?

I think the odo supported features is mostly used to indicate that we can run the command "odo dev/deploy". Even if the component has been deployed, doesn't mean we can run it now (in this case, we couldn't, because we don't have the Devfile)

**Comment by rm3l** [comment]
> > * Also, I understood that the "Supported odo features" information is determined from the current Devfile and currently displays "Unknown" when running with `--name`. When running with the `--name` flag, if a mode (`Dev` and/or `Deploy`) is detected and listed in the "Running in" section, it means that this feature is supported by `odo`, no?
> 
> I think the odo supported features are mostly intended to indicate that we can run the command "odo dev/deploy". Even if the component has been deployed, doesn't mean we can run it now (in this case, we couldn't, because we don't have the Devfile)

Okay, I see. Thanks.

**Pull request merged**